tls/s2n_client_hello.c

```diff
@@ -433,9 +433,27 @@ int s2n_process_client_hello(struct s2n_connection *conn)
     /* Find potential certificate matches before we choose the cipher. */
     POSIX_GUARD(s2n_conn_find_name_matching_certs(conn));
 
+    /* Save the previous cipher suite */
+    uint8_t previous_cipher_suite_iana[S2N_TLS_CIPHER_SUITE_LEN] = { 0 };
+    POSIX_CHECKED_MEMCPY(previous_cipher_suite_iana, conn->secure.cipher_suite->iana_value, S2N_TLS_CIPHER_SUITE_LEN);
+
     /* Now choose the ciphers we have certs for. */
     POSIX_GUARD(s2n_set_cipher_as_tls_server(conn, client_hello->cipher_suites.data, client_hello->cipher_suites.size / 2));
 
+    /**
+     *= https://tools.ietf.org/rfc/rfc8446#4.1.4
+     *# Servers MUST ensure that they negotiate the
+     *# same cipher suite when receiving a conformant updated ClientHello (if
+     *# the server selects the cipher suite as the first step in the
+     *# negotiation, then this will happen automatically).
+     **/
+    uint8_t null_cipher[S2N_TLS_CIPHER_SUITE_LEN] = {TLS_NULL_WITH_NULL_NULL};
+    if (s2n_is_hello_retry_handshake(conn) &&
+        !s2n_constant_time_equals(previous_cipher_suite_iana, null_cipher, S2N_TLS_CIPHER_SUITE_LEN)) {
```

Comment:
I'm not sure using `s2n_is_hello_retry_message` would work here (unless I'm misunderstanding something). In this check, the server wants to know if this is the first client hello or the second client hello. I think `s2n_is_hello_retry_message` wouldn't tell us this, right? I think checking s2n_conn_get_current_message_type won't work either, since it will return a client hello in both cases.

One way that might work would be to check if the message number is > 0, meaning it's not the first client hello. Would that be better? Or, I may be misunderstanding your suggestion of using `s2n_is_hello_retry_message`.

Reply:
I switched this with a check that uses the message number instead: 6a5ced03490d0e6336a478e948bb4b8403d202f5

Is this reasonable?